src/QueryPipeline/QueryPipelineBuilder.cpp

```diff
@@ -618,6 +619,31 @@ void QueryPipelineBuilder::setProgressCallback(ProgressCallback callback)
     progress_callback = callback;
 }
 
+void QueryPipelineBuilder::connectDependencies()
+{
+    std::vector<RemoteSource *> input_dependencies;
+    std::vector<ReadFromMergeTreeDependencyTransform *> output_dependencies;
+
+    for (auto & processor : *pipe.getProcessorsPtr())
+    {
+        if (auto * remote_dependency = dynamic_cast<RemoteSource *>(processor.get()); remote_dependency)
+            input_dependencies.emplace_back(remote_dependency);
+        if (auto * merge_tree_dependency = dynamic_cast<ReadFromMergeTreeDependencyTransform *>(processor.get()); merge_tree_dependency)
+            output_dependencies.emplace_back(merge_tree_dependency);
+    }
+
+    if (input_dependencies.empty() || output_dependencies.empty())
```

Comment:
Maybe also after check , additionally check that both are not empty, just in case

Reply:
This is not necessary as I understand. After this check we are sure that they are both not empty, because otherwise...